modules/primo/src/Primo/Ting/Result.php

```diff
@@ -133,8 +143,35 @@ public function getSearchRequest() {
    *   List of facets, empty if none where found.
    */
   public function getFacets() {
-    // TODO: Implement.
-    return [];
+    // Return the mapped facets if we've already done the work.
+    if ($this->facets !== NULL) {
+      return $this->facets;
+    }
+
+    // Prepare for loading.
+    $this->facets = [];
+
+    // Handle empty facets.
+    $primo_facets = $this->result->getFacets();
+    if (empty($primo_facets)) {
+      return $this->facets;
+    }
+
+    // Convert Primo Facets to TingSearchFacetTerm instances.
+    array_walk($primo_facets, function (Facet $facet) {
+      $terms = [];
+      $values = $facet->getValues();
+      array_walk($values, function ($frequency, $name) use (&$terms) {
+        $terms[] = new TingSearchFacetTerm($name, $frequency);
+      });
```

Comment:
I just discovered that [you can pass multiple arrays to `array_map()` which is transferred to multiple arguments to the callback](http://php.net/manual/en/function.array-map.php#81767). This makes the following possible:

```php
$terms = array_map(function($name, $frequency) {
  return new TingSearchFacetTerm($name, $frequency);
}, array_keys($values), $values);
```

Reply:
nice!